File: addtwonums.py
# Definition for singly-linked list.
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def makeNum(self,l1:ListNode)->int:
        base = 0
        number = 0
        while l1:
            number += pow(10,base) * l1.val
            base += 1
            l1 = l1.next
        
        return number

    def makeList(self,num:int)->ListNode:
        digit = num%10
        num //= 10
        sumNode = ListNode(digit)
        copy = sumNode
        
        while num!=0:
            digit = num%10
            num //= 10
            newNode = ListNode(digit)
            sumNode.next = newNode
            sumNode = sumNode.next
        
        return copy

    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        return self.makeList(self.makeNum(l1)+self.makeNum(l2))
        

File: addtwonums.cpp
struct ListNode {
         int val;
         ListNode *next;
         ListNode() : val(0), next(nullptr) {}
         ListNode(int x) : val(x), next(nullptr) {}
         ListNode(int x, ListNode *next) : val(x), next(next) {}
     };

class Solution {
    public:
        unsigned long long power(int exp){
            unsigned long long base = 1;
            for(int i{1};i<=exp;i++){
                base *= 10;
            }

            return base;
        }

        unsigned long long makeNum(ListNode* list){
            ListNode* copy = list;
            int base = 0;
            unsigned long long number = 0;
            while(copy){
                number += power(base) * copy->val;
                base += 1;
                copy = copy->next;
            }

            return number;
        }

        ListNode* makeList(unsigned long long num){
            int digit = num%10;
            num /= 10;
            ListNode* sumNode = new ListNode(digit);
            ListNode* copy = sumNode;

            while(num!=0){
                digit = num%10;
                num /= 10;
                ListNode* newNode = new ListNode(digit);
                sumNode->next = newNode;
                sumNode = sumNode->next;
            }

            return copy;
        }

        ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
            return makeList(makeNum(l1)+makeNum(l2));
        }
    };
    